.tools/bin/check-i18n

```diff
@@ -0,0 +1,112 @@
+#!/usr/bin/env php
+<?php
+
+/**
+ * @author Thomas Blum
+ *
+ * This command checks for removable keys in i18n lamg files.
+ *
+ * Examples:
+ *
+ * bin/check-i18n
+ * bin/check-i18n core
+ * bin/check-i18n -addonName-
+ */
+
+unset($REX);
+$REX['REDAXO'] = true;
+$REX['HTDOCS_PATH'] = './';
+$REX['BACKEND_FOLDER'] = 'redaxo';
+$REX['LOAD_PAGE'] = false;
+
+require $REX['BACKEND_FOLDER'] . '/src/core/boot.php';
+require $REX['BACKEND_FOLDER'] . '/src/core/packages.php';
+
+// avoid PREG_JIT_STACKLIMIT_ERROR, changelog parts can be long
+ini_set('pcre.jit', 0);
+
+/**
+ * @package redaxo\core
+ *
+ * @internal
+ */
+class rex_check_i18n
+{
+    private const PATTERN = '
+        @
+        (?<complete>
+            \s*
+            (?>rex_i18n::|->)(?>msg|msgInLocale|rawMsg|rawMsgInLocale|hasMsg|hasMsgOrFallback|addMsg|translate|translateArray)\(
+                \s*
+                (?>\'|")
+                (?<key>.*?)
+                (?>\'|")
+                .*?
+            \)
+        )
+        @isx';
+
+    private $dir;
+    private $iterator;
+    private $i18nInUse;
+    private $i18nMsg;
+
+    public function __construct($dir)
+    {
+        $this->dir = $dir;
+        $this->iterator = rex_finder::factory($this->dir)->recursive()->filesOnly()->ignoreFiles(['.*']);
+    }
+
+    private function parse()
+    {
+        /** @var \SplFileInfo $file */
+        foreach ($this->iterator as $file) {
+            $filepath = $file->getPathname();
+            $content = \rex_file::get($filepath);
+
+            if ('lang' === $file->getExtension() && preg_match_all('/^([^=\s]+)\h*=\h*(\S.*)(?<=\S)/m', $content, $matches, PREG_SET_ORDER | PREG_OFFSET_CAPTURE)) {
```

Comment:
ggf könnten wir die rex_i18n klasse verwenden um die files zu parsen?

Reply:
Die Dateien werden dort schon geparsed. Es fehlt nur eine Methode um an diese Sprachkeys heranzukommen.

Außerdem lasse ich den Pfad zur Datei und die Zeilennummern mit ausgeben, was die Klasse momentan selbst nicht speichert.

Ich lade morgen mal einen Screenshot hoch.

Reply:
<img width="484" alt="Bildschirmfoto 2020-06-24 um 08 08 38" src="https://user-images.githubusercontent.com/338267/85507050-f3305c00-b5f1-11ea-8660-842934f46edb.png">

Die jetzige Ausgabe soll nicht so bleiben. Vielleicht könnten wir uns vorher auf eine Richtung einigen. Was würdet ihr erwarten?

Reply:
Das ausgabeformat ist mir nicht so wichtig. Kannst du anhand deines formates bereits für die core lang dateien die zu löschenden keys in einem separaten PR posten, sodass wir sehen können ob es noch false positives gibt o.ä.?

Reply:
Anders gesagt: ich denke wir brauch kein ausgabeformat, sondern ändern direkt die lang dateien und fertig, oder?

Reply:
> Kannst du anhand deines formates bereits für die core lang dateien die zu löschenden keys in einem separaten PR posten (also die nach dem skript gefixten dateien), sodass wir sehen können ob es noch false positives gibt bzw.

Habe hier einmal einen PR gemacht
https://github.com/redaxo/redaxo/pull/3767

Wichtig zu wissen ist, dass eben nur in `core` abgeglichen wird. Wird ein Key nur in AddOns verwendet und nicht im Core  (Bsp. `add`), dann wird dieser aktuell ebenso aus den Core Dateien gelöscht.

Reply:
> Wichtig zu wissen ist, dass eben nur in `core` abgeglichen wird. Wird ein Key nur in AddOns verwendet und nicht im Core (Bsp. `add`), dann wird dieser aktuell ebenso aus den Core Dateien gelöscht.

ok, ich denke für solche keys, die sehr allgemein sind und von denen wir wissen dass sie übergreifend verwendet werden können wir im tool ggf. eine whitelist hinterlegen

Reply:
Wobei ich da eigentlich mehr hin möchte, dass Addons nicht immer wieder die selben Übersetzungen hinzufügen (mit eigenem Präfix), sondern dass allgemeine Dinge besser nur einmal im Core sind, und die Addons die dann nutzen können.

Reply:
> Wobei ich da eigentlich mehr hin möchte, dass Addons nicht immer wieder die selben Übersetzungen hinzufügen (mit eigenem Präfix), sondern dass allgemeine Dinge besser nur einmal im Core sind, und die Addons die dann nutzen können.

hmm bin nicht sicher wie ich das deuten soll im aktuellen context. ich habe nicht sagen wollen dass wir keys in addons duplizieren sollten

Reply:
Vielleicht legen wir im Core 2 Sprachfiles an. Eine wo nur Keys enthalten sind, die der Core selber nutzt und eine für allgemeine Keys die von alle verwendet werden können. Damit hätten wir dann automatisch ein Whitelist-File was immer unberührt bleibt.

Reply:
Ich wollte damit aussagen, dass ich eher nicht mit einer expliziten Liste arbeiten würde, da mein Ziel wäre, das es in Zukunft eher mehr Fälle werden als weniger.
(Keys im Core, die nur von Addons verwendet werden)